blueprint: add cacert customization

lzap opened the pull request:
This is a follow-up to 

https://github.com/osbuild/images/pull/907/files

I am shooting in the dark, not sure how can I test this locally. Let me know how can I test this.

bcl commented on internal/blueprint/ca_customizations.go:
```diff
@@ -0,0 +1,5 @@
+package blueprint
+
+type CACustomization struct {
```
I think this belongs in the customizations.go with the rest of the structs. Putting it into a separate file makes it harder to read the code.

mvo5 commented on internal/blueprint/ca_customizations.go:
```diff
@@ -0,0 +1,5 @@
+package blueprint
+
+type CACustomization struct {
```
This maybe out of scope for this PR but we have exactly this struct define in "images" and it is public, couldn't we just import it instead of copying it (the same question for most of the others but I guess this is why it's out of scope).

lzap replied:
I have no idea why is it like that, I am just following the pattern here. I am running a research parallel to this to find a common blueprint that should solve this problem for once. I would say let’s not change this in this PR?

avitova reviewed (commented):
Few nitpicks I noticed.

lzap commented on vendor/github.com/osbuild/images/pkg/blueprint/customizations.go:
```diff
@@ -31,6 +33,7 @@ type Customizations struct {
 	Installer          *InstallerCustomization        `json:"installer,omitempty" toml:"installer,omitempty"`
 	RPM                *RPMCustomization              `json:"rpm,omitempty" toml:"rpm,omitempty"`
 	RHSM               *RHSMCustomization             `json:"rhsm,omitempty" toml:"rhsm,omitempty"`
+	CACerts            *CACustomization               `json:"cacerts,omitempty" toml:"ca,omitempty"`
```
Ah thanks, here as well: https://github.com/osbuild/images/pull/1076

lzap commented on vendor/github.com/osbuild/images/pkg/blueprint/customizations.go:
```diff
@@ -425,3 +435,32 @@ func (c *Customizations) GetRHSM() *RHSMCustomization {
+		return nil
+	}
+
+	if c.CACerts != nil {
+		for _, bundle := range c.CACerts.PEMCerts {
+			_, err := cert.ParseCerts(bundle)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
```
This is a vendored file, feel free to file a patch into `images` project. Thanks!

lzap replied:
Will address that in the separate PR in the images project.

avitova replied:
Amazing!:)

lzap commented on vendor/github.com/osbuild/images/pkg/blueprint/filesystem_customizations.go:
```diff
@@ -93,3 +71,27 @@ func CheckMountpointsPolicy(mountpoints []FilesystemCustomization, mountpointAll
+	switch s := size.(type) {
+	case string:
+		return datasizes.Parse(s)
+	case int64:
+		if s < 0 {
+			return 0, fmt.Errorf("cannot be negative")
+		}
+		return uint64(s), nil
+	case float64:
+		if s < 0 {
+			return 0, fmt.Errorf("cannot be negative")
+		}
+		// TODO: emit warning of possible truncation?
+		return uint64(s), nil
```
This is a vendored file, feel free to file a patch into `images` project. Thanks!

lzap replied:
This does not compile on Go 1.22 locally: `invalid operation: s < 0 (operator < not defined on interface)` so skipping. This is also not my code, went in via images library update it was a different commit.

lzap commented:
Rebased, how about this? Anything else that needs to be done?

thozza reviewed (commented):
LGTM.

Could you please extend the Cloud API `api.sh` test case and at least one Weldr API test case (e.g. `aws.sh`) to use this customization? You can take a look at the following commit as an inspiration https://github.com/osbuild/osbuild-composer/commit/768537dba3730cc70ad7a00fc0f868f88692ced8#diff-8c9fc1f11806f990402d945e7a6a3bdcf8be429120a4b8c2ec18e9ff5251f02f , but also extend the `_instanceCheck()` in https://github.com/osbuild/osbuild-composer/blob/main/test/cases/api/common/common.sh#L18

avitova commented:
Thanks for the other PR, I am good here.

lzap commented:
I see four people commenting, no formal review tho. Please give me the green light, enabled auto merge, thanks.

thozza reviewed (changes requested):
> I see four people commenting, no formal review tho. Please give me the green light, enabled auto merge, thanks.

Consider [my comment](https://github.com/osbuild/osbuild-composer/pull/4487#pullrequestreview-2475082064) to be a formal review. I usually do not explicitly request request changes, because it feels a bit rough, compared to just making a comment.

lzap commented:
Rebased, how about this.

avitova reviewed (commented):
I am good here, but let's wait for others to approve too. :)

lzap commented:
I changed the test case to `grep` the concatenated bundle instead of trying to find the cert copy in `directory-hash`. I have learned that this directory is not present on older RHEL systems so this way the test will pass everywhere. Rebased.

thozza reviewed (dismissed):
LGTM.
Thank you!

thozza commented:
@lzap you need to append `${CACERTS_CUSTOMIZATION_BLOCK}` also in:

- https://github.com/osbuild/osbuild-composer/blob/main/test/cases/api/gcp.sh#L75
- https://github.com/osbuild/osbuild-composer/blob/main/test/cases/api/oci.sh#L80

lzap commented:
RIght, added, rebased, no changes. Thanks for help with both! :-D
